syntax = "proto3";

import "google/protobuf/any.proto";

package protocol;

// https://github.com/tronprotocol/protocol/blob/2a678934da3992b1a67f975769bbb2d31989451f/core/Tron.proto#L336
message Transaction {
    message Contract {
        enum ContractType {
            AccountCreateContract = 0;
            TransferContract = 1;
            TransferAssetContract = 2;
            VoteAssetContract = 3;
            VoteWitnessContract = 4;
            FreezeBalanceContract = 11;
            UnfreezeBalanceContract = 12;
            WithdrawBalanceContract = 13;
            UnfreezeAssetContract = 14;
            TriggerSmartContract = 31;
            FreezeBalanceV2Contract = 54;
            UnfreezeBalanceV2Contract = 55;
            WithdrawExpireUnfreezeContract = 56;
            DelegateResourceContract = 57;
            UnDelegateResourceContract = 58;
        }
        ContractType type = 1;
        google.protobuf.Any parameter = 2;
        bytes provider = 3;
    }

    message raw {
        bytes ref_block_bytes = 1;
        int64 ref_block_num = 3;
        bytes ref_block_hash = 4;
        int64 expiration = 8;
        // transaction memo
        bytes data = 10;
        //only support size = 1,  repeated list here for extension
        repeated Contract contract = 11;
        int64 timestamp = 14;
        int64 fee_limit = 18;
    }

    raw raw_data = 1;
}

message BlockHeader {
    message raw {
        int64 timestamp = 1;
        bytes tx_trie_root = 2;
        bytes parent_hash = 3;
        int64 number = 7;
        int64 witness_id = 8;
        bytes witness_address = 9;
        int32 version = 10;
    }
    raw raw_data = 1;
    bytes witness_signature = 2;
}

message TransferContract {
    bytes owner_address = 1;
    bytes to_address = 2;
    int64 amount = 3;
}

message TransferAssetContract {
    bytes asset_name = 1;
    bytes owner_address = 2;
    bytes to_address = 3;
    int64 amount = 4;
}

enum ResourceCode {
    BANDWIDTH = 0x00;
    ENERGY = 0x01;
}

message FreezeBalanceContract {
    bytes owner_address = 1;
    int64 frozen_balance = 2;
    int64 frozen_duration = 3;

    ResourceCode resource = 10;
    bytes receiver_address = 15;
}

message FreezeBalanceV2Contract {
    bytes owner_address = 1;
    int64 frozen_balance = 2;
    ResourceCode resource = 3;
}

message UnfreezeBalanceContract {
    bytes owner_address = 1;

    ResourceCode resource = 10;
    bytes receiver_address = 15;
}

message UnfreezeBalanceV2Contract {
    bytes owner_address = 1;
    int64 unfreeze_balance = 2;
    ResourceCode resource = 3;
}

message WithdrawExpireUnfreezeContract {
    bytes owner_address = 1;
}

message DelegateResourceContract {
    bytes owner_address = 1;
    ResourceCode resource = 2;
    int64 balance = 3;
    bytes receiver_address = 4;
    bool  lock = 5;
}

message UnDelegateResourceContract {
    bytes owner_address = 1;
    ResourceCode resource = 2;
    int64 balance = 3;
    bytes receiver_address = 4;
}

message UnfreezeAssetContract {
    bytes owner_address = 1;
}

message VoteAssetContract {
    bytes owner_address = 1;
    repeated bytes vote_address = 2;
    bool support = 3;
    int32 count = 5;
}

message VoteWitnessContract {
    message Vote {
        bytes vote_address = 1;
        int64 vote_count = 2;
    }
    bytes owner_address = 1;
    repeated Vote votes = 2;
    bool support = 3;
}

message WithdrawBalanceContract {
    bytes owner_address = 1;
}

message TriggerSmartContract {
    bytes owner_address = 1;
    bytes contract_address = 2;
    int64 call_value = 3;
    bytes data = 4;
    int64 call_token_value = 5;
    int64 token_id = 6;
}
